### ISSUE TITLE
Bug with tcp_nodelay option

### ISSUE THREAD

**Issue by ledergec** [eclipse/mosquitto#962]
Using mosquitto 1.5.1:

When we set the tcp_nodelay option to true in the config we kept getting the following Warning:
"Warning: Unable to set TCP_NODELAY."

This was consistent in docker and our yocto embedded environment.

When debuggin mosquitto we realized that the setsockopt syscall set errno to EINVAL which indicates a bug in mosquitto.

The patch below fixes the issue for us:

```
diff --git a/src/net.c b/src/net.c
index 64f8bdc..6b4d717 100644
--- a/src/net.c
+++ b/src/net.c
@@ -104,6 +104,13 @@ int net__socket_accept(struct mosquitto_db *db, mosq_sock_t listensock)
 	if(net__socket_nonblock(&new_sock)){
 		return INVALID_SOCKET;
 	}
+	
+	if(db->config->set_tcp_nodelay){
+		int flag = 1;
+		if(setsockopt(new_sock, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int)) != 0){
+			log__printf(NULL, MOSQ_LOG_WARNING, "Warning: Unable to set TCP_NODELAY.");
+		}
+	}
 
 #ifdef WITH_WRAP
 	/* Use tcpd / libwrap to determine whether a connection is allowed. */
@@ -119,13 +126,6 @@ int net__socket_accept(struct mosquitto_db *db, mosq_sock_t listensock)
 	}
 #endif
 
-	if(db->config->set_tcp_nodelay){
-		int flag = 1;
-		if(setsockopt(new_sock, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(int) != 0)){
-			log__printf(NULL, MOSQ_LOG_WARNING, "Warning: Unable to set TCP_NODELAY.");
-		}
-	}
-
 	new_context = context__init(db, new_sock);
 	if(!new_context){
 		COMPAT_CLOSE(new_sock);
```

**Comment by kumar003vinod** [comment]
Hi @ledergec ,

A fix for this is done with issue #941 .

**Comment by ralight** [comment]
@ledergec thanks for taking the time to figure out the problem and report it. As @kumar003vinod says, this is already fixed and will be released shortly.
